### PR TITLE
Remove some .NET specific logic from serialization/deserialization

### PR DESCRIPTION
**Pull request by JoshLove-msft**
There was some .NET specific logic where we were attempting to only store sanitized request bodies as JSON if the serialization would leave the bytes unchanged. If reserializing would have reformatted/re-encoded the body, we instead serialized as string. This could happen if characters such as & were included in the body. This behavior made recordings inconsistent and harder to read. Instead of doing this, we will always serialize as JSON, but use the STJ RelaxedJsonEncoding, to prevent excessive escaping of special characters.

The PR also always re-normalizes the sanitized request before matching using STJ RelaxedJsonEncoding, as well as when loading the requests from stored session records. This prevents matching issues that could arise from reformatting that occurs while sanitizing. Now we are always reformatting in a consistent way.

This change is an alternative to the new setting proposed in https://github.com/Azure/azure-sdk-tools/pull/2427

### CODE REVIEW

File: tools/test-proxy/Azure.Sdk.Tools.TestProxy.Tests/RecordSessionTests.cs [hunk 1]
```diff
@@ -0,0 +1,572 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (JoshLove-msft) [inline tools/test-proxy/Azure.Sdk.Tools.TestProxy.Tests/RecordSessionTests.cs:1]: Copied this over from .NET repo.

File: tools/test-proxy/Azure.Sdk.Tools.TestProxy/Common/RecordEntry.cs [hunk 2]
```diff
@@ -92,29 +91,50 @@ private static byte[] DeserializeBody(IDictionary<string, string[]> headers, in
...
-                {
-                    StringBuilder stringBuilder = new StringBuilder();
+                using var memoryStream = new MemoryStream();
+                using var writer = new Utf8JsonWriter(memoryStream, WriterOptions);
 
-                    foreach (JsonElement item in property.EnumerateArray())
+                switch (property.ValueKind)
+                {
+                    case JsonValueKind.Object:
+                        property.WriteTo(writer);
```
Comment (pakrym) [inline tools/test-proxy/Azure.Sdk.Tools.TestProxy/Common/RecordEntry.cs:100]: I'd rewrite this method so it extracts byte[] first and then tried to reformat it if it's JSON.

RIght now all the fallbacks are hard to follow.

File: tools/test-proxy/Azure.Sdk.Tools.TestProxy/Common/RecordSession.cs [hunk 3]
```diff
@@ -101,6 +102,21 @@ public RecordEntry Lookup(RecordEntry requestEntry, RecordMatcher matcher, IEnum
                 sanitizer.Sanitize(requestEntry);
             }
 
+            try
```
Comment (pakrym) [inline tools/test-proxy/Azure.Sdk.Tools.TestProxy/Common/RecordSession.cs:105]: Extract into a method and use for record loading.

**Comment by pakrym** [comment]
Let's add a test that ensures that re-normalization works when matching and loading.

**Comment by pakrym** [comment]
Can you also check if this solves the number formatting problem and add a test if it does?

**Comment by pakrym** [comment]
Do we need to update content-length values after normalization, especially during the record loading?

File: tools/test-proxy/Azure.Sdk.Tools.TestProxy/Common/RecordEntry.cs [hunk 4]
```diff
@@ -156,15 +178,15 @@ public void Serialize(Utf8JsonWriter jsonWriter)
             SerializeHeaders(jsonWriter, Request.Headers);
             jsonWriter.WriteEndObject();
 
-            SerializeBody(jsonWriter, "RequestBody", Request.Body, Request.Headers, RequestWriterOptions);
+            SerializeBody(jsonWriter, "RequestBody", Request.Body, Request.Headers, WriterOptions);
```
Comment (pakrym) [inline tools/test-proxy/Azure.Sdk.Tools.TestProxy/Common/RecordEntry.cs:181]: Do we need the parameter?

File: tools/test-proxy/Azure.Sdk.Tools.TestProxy/Common/RecordSession.cs [hunk 5]
```diff
@@ -102,6 +95,19 @@ public RecordEntry Lookup(RecordEntry requestEntry, RecordMatcher matcher, IEnum
...
             }
+            return LookUpSanitizedRecord(requestEntry, matcher, remove);
+        }
+
+        private RecordEntry LookUpSanitizedRecord(RecordEntry requestEntry, RecordMatcher matcher, bool remove)
+        {
+            try
+            {
+                // normalize request body with STJ using relaxed escaping to match behavior when Deserializing from session files
+                RecordEntry.NormalizeJsonBody(requestEntry.Request, requestEntry.Request.Body);
```
Comment (pakrym) [inline tools/test-proxy/Azure.Sdk.Tools.TestProxy/Common/RecordSession.cs:106]: We should do this at the place that builds the RecordEntry.

File: tools/test-proxy/Azure.Sdk.Tools.TestProxy/Common/RecordSession.cs [hunk 6]
```diff
@@ -102,6 +95,19 @@ public RecordEntry Lookup(RecordEntry requestEntry, RecordMatcher matcher, IEnum
...
+            }
+            catch (JsonException)
+            {
+            }
```
Comment (scbedd) [inline tools/test-proxy/Azure.Sdk.Tools.TestProxy/Common/RecordSession.cs:110]: Nit: we explicitly catch `JsonException` in `NormalizeJsonBody`. It's not possible that we actually get this exception here. Perhaps swap to `Exception` and throw an `HttpException` if we ever seen it?

**Review by scbedd — APPROVED** [review]
Nothing jumps out at me as weird. Appreciate the update!

File: tools/test-proxy/Azure.Sdk.Tools.TestProxy/Common/RecordSession.cs (hunk 6 above)
Comment (JoshLove-msft) [reply tools/test-proxy/Azure.Sdk.Tools.TestProxy/Common/RecordSession.cs:110]: Ah good catch - will just remove it from here.

File: tools/test-proxy/Azure.Sdk.Tools.TestProxy/Common/RecordSession.cs [hunk 7]
```diff
@@ -102,6 +95,19 @@ public RecordEntry Lookup(RecordEntry requestEntry, RecordMatcher matcher, IEnum
             {
                 sanitizer.Sanitize(requestEntry);
             }
+            return LookUpSanitizedRecord(requestEntry, matcher, remove);
```
Comment (pakrym) [inline tools/test-proxy/Azure.Sdk.Tools.TestProxy/Common/RecordSession.cs:98]: Merge methods?

Comment (pakrym) [reply tools/test-proxy/Azure.Sdk.Tools.TestProxy/Common/RecordSession.cs:98]: Now that normalization is a single line.

File: tools/test-proxy/Azure.Sdk.Tools.TestProxy/Common/RecordEntry.cs [hunk 8]
```diff
@@ -77,52 +76,75 @@ public static RecordEntry Deserialize(JsonElement element)
...
-            if (property.ValueKind == JsonValueKind.Array)
+            else
             {
-                return Array.Empty<byte>();
+                requestOrResponse.Body = Convert.FromBase64String(property.GetString());
             }
+        }
 
-            return Convert.FromBase64String(property.GetString());
+        public static void NormalizeJsonBody(RequestOrResponse requestOrResponse, byte[] bytes)
```
Comment (pakrym) [inline tools/test-proxy/Azure.Sdk.Tools.TestProxy/Common/RecordEntry.cs:131]: Why do you pass bytes in?

Comment (pakrym) [reply tools/test-proxy/Azure.Sdk.Tools.TestProxy/Common/RecordEntry.cs:131]: Can we use only the RequestOrResponse?

File: tools/test-proxy/Azure.Sdk.Tools.TestProxy/Common/RecordSession.cs (hunk 5 above)
Comment (JoshLove-msft) [reply tools/test-proxy/Azure.Sdk.Tools.TestProxy/Common/RecordSession.cs:106]: this is the place where the sanitization occurs so it makes sense to keep here.

File: tools/test-proxy/Azure.Sdk.Tools.TestProxy/Common/RecordEntry.cs (hunk 8 above)
Comment (JoshLove-msft) [reply tools/test-proxy/Azure.Sdk.Tools.TestProxy/Common/RecordEntry.cs:131]: when deserializing the body will not be populated on the requestOrResponse

File: tools/test-proxy/Azure.Sdk.Tools.TestProxy/Common/RecordEntry.cs [hunk 9]
```diff
@@ -14,18 +14,17 @@ namespace Azure.Sdk.Tools.TestProxy.Common
 {
     public class RecordEntry
     {
-        private static readonly JsonWriterOptions RequestWriterOptions = new JsonWriterOptions();
-        // Responses are usually formatted using Newtonsoft.Json that has more relaxed encoding rules
+        // Requests and responses are usually formatted using Newtonsoft.Json that has more relaxed encoding rules
         // To enable us to store more responses as JSON instead of string in Recording files use
         // relaxed settings for roundtrip
-        private static readonly JsonWriterOptions ResponseWriterOptions = new JsonWriterOptions()
+        public static readonly JsonWriterOptions WriterOptions = new JsonWriterOptions()
```
Comment (pakrym) [inline tools/test-proxy/Azure.Sdk.Tools.TestProxy/Common/RecordEntry.cs:20]: Who else is using this property?

File: tools/test-proxy/Azure.Sdk.Tools.TestProxy/Common/RecordEntry.cs (hunk 8 above)
Comment (pakrym) [reply tools/test-proxy/Azure.Sdk.Tools.TestProxy/Common/RecordEntry.cs:131]: So what about populating it right before calling the method?

Comment (JoshLove-msft) [reply tools/test-proxy/Azure.Sdk.Tools.TestProxy/Common/RecordEntry.cs:131]: I suppose I can just set it.

File: tools/test-proxy/Azure.Sdk.Tools.TestProxy/Common/RecordSession.cs (hunk 7 above)
Comment (JoshLove-msft) [reply tools/test-proxy/Azure.Sdk.Tools.TestProxy/Common/RecordSession.cs:98]: they have different signatures, I suppose we can just delete PlaybackTransport so we only would need the one signature.

**Review by pakrym — APPROVED** [review]
